osu.Game/Screens/Edit/Timing/MetronomeDisplay.cs

```diff
@@ -0,0 +1,270 @@
+                            Anchor = Anchor.BottomCentre,
+                            Origin = Anchor.BottomCentre,
+                            Size = triangleSize,
+                            Colour = overlayColourProvider.Background2,
+                            EdgeSmoothness = new Vector2(1),
+                            Alpha = 0.8f
+                        },
+                    }
+                },
+                bpmText = new OsuSpriteText
+                {
+                    Name = @"BPM display",
+                    Colour = overlayColourProvider.Content1,
+                    Anchor = Anchor.BottomCentre,
+                    Origin = Anchor.BottomCentre,
+                    Y = -3,
+                },
+            };
+        }
+
+        private double beatLength;
+
+        private TimingControlPoint timingPoint;
+
+        private bool isSwinging;
+
+        private readonly BindableInt interpolatedBpm = new BindableInt();
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            interpolatedBpm.BindValueChanged(bpm => bpmText.Text = bpm.NewValue.ToString(CultureInfo.CurrentCulture));
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            timingPoint = editorBeatmap.ControlPointInfo.TimingPointAt(BeatSyncClock.CurrentTime);
+
+            if (beatLength != timingPoint.BeatLength)
+            {
+                beatLength = timingPoint.BeatLength;
+
+                EarlyActivationMilliseconds = timingPoint.BeatLength / 2;
+
+                float bpmRatio = (float)Interpolation.ApplyEasing(Easing.OutQuad, Math.Clamp((timingPoint.BPM - 30) / 480, 0, 1));
+
+                weight.MoveToY((float)Interpolation.Lerp(0.1f, 0.83f, bpmRatio), 600, Easing.OutQuint);
+                this.TransformBindableTo(interpolatedBpm, (int)Math.Round(timingPoint.BPM), 600, Easing.OutQuint);
+            }
+
+            if (BeatSyncClock?.IsRunning != true && isSwinging)
+            {
+                swing.ClearTransforms(true);
+
+                using (swing.BeginDelayedSequence(350))
+                {
+                    swing.RotateTo(0, 1000, Easing.OutQuint);
+                    stick.FadeColour(overlayColourProvider.Colour2, 1000, Easing.OutQuint);
+                }
+
+                isSwinging = false;
+            }
+        }
+
+        protected override void OnNewBeat(int beatIndex, TimingControlPoint timingPoint, EffectControlPoint effectPoint, ChannelAmplitudes amplitudes)
+        {
+            base.OnNewBeat(beatIndex, timingPoint, effectPoint, amplitudes);
+
+            const float angle = 27.5f;
+
+            if (!IsBeatSyncedWithTrack)
+                return;
+
+            isSwinging = true;
+
+            float currentAngle = swing.Rotation;
+            float targetAngle = currentAngle > 0 ? -angle : angle;
+
+            swing.RotateTo(targetAngle, beatLength, Easing.InOutQuad);
```

Comment:
This is going to be a bit of an insane point but if we're implementing a metronome display in the game then I'll bring it up - for physical correctness I believe this should be

```suggestion
            swing.RotateTo(targetAngle, beatLength, Easing.InOutSine);
```

if I remember my oscillator physics right. (Note that the weight placement is probably not all there either but I'm not nitpicking on that since (a) it looks close enough to a real metronome already with the non-linear placement of the weight and, (b) I'm not really sure what is correct - I think it'd be something or other to do with the inverse of the squared frequency value?)

Reply:
I'll check again, but Quad looked more correct to me. I originally did have it on sine.

Reply:
`Quad` still looks more correct to my eye. I'm willing to concede if someone does a comparison to a physical metronome. `Sine` just doesn't feel like what I've come to expect from a metronome swing (not sure `Quad` is perfect either, but feels closer).

Reply:
`Sine` is [small-angle approximation](https://en.wikipedia.org/wiki/Pendulum_(mechanics)#Small-angle_approximation). There is no closed-formula for the exact parametric representation.